### ISSUE TITLE
Notebook Controller should surface StatefulSet status in the Notebook CR if Pod is not created

### ISSUE THREAD

**Issue by kimwnasptd** [kubeflow/kubeflow#3772]
The Notebook Controller sets the `status.containerState` in the Notebook CR only when the Pod is created. So, if there would be a problem with a missing Service Account or an unbound PVC, then the Pod will not be scheduled and thus the status in the Notebook CR will not be updated to provide the relative info.

The Controller should set the status in the Notebook CR even when the Pod is not scheduled. In this case, the Notebook Controller will need to query the [Events](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.12/#event-v1-core) API and select the ones with
```yaml
involvedObject:
    apiVersion: apps/v1
    kind: StatefulSet
    name: <Notebook Name>
    namespace: <Notebook Namespace>
```

From the selected events, it should then only keep the latest one. From this event we need the `.message` field.

Lastly, in order for the jupyter-web-app to show a spinner and the status, the Notebook Controller will need to set the `status.containerState` in the CR to `waiting` and the `waiting.reason` to be the StatefulSet's event message.

/cc @lluunn

**Comment by fediazgon** [comment]
Is this still an issue. Can I work on it?

**Comment by jlewi** [comment]
@fediazgon Yes this is still an issue; it would be great if you could pick it up.

**Comment by jlewi** [comment]
@fediazgon Do you plan on picking this up? If yes do you have an ETA for when you might have an initial PR that people can take a look at?

**Comment by fediazgon** [comment]
@jlewi Yes! First time diving in k8s code and I'm derailing a lot because I'm trying to understand the surrounding code. I'll try to have a PR before Saturday.

**Comment by fediazgon** [comment]
Hey @jlewi @kimwnasptd. Before start coding, I would like some advice to see if the following approach is valid.

Events related to a notebook are named like `notebookName.XXXXXXX`, this means that, if we watch for single Events, we can only capture the event at the beginning of the reconcile call, since later (e.g., after we fetch the notebook instance) we don't know the name of the event. This problem does not exist with pods because we know pods are named like `notebookName-0`. However, for events, the `XXXXXXX` is randomly generated. One possible implementation is:

1) Split the event's name by ".", get the first substring (i.e., "notebookName"). Try to fetch the notebook by using key `types.NamespacedName{Name: substring, Namespace: event.Namespace}`. The previous might fail because the event could be unrelated to a notebook. If a notebook was retrieved, update its state if necessary. 

This relies on events' names being like "eventSource.XXXXXXX". I could not find anything in [kubernetes - working with names](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names) about these names being consistent.

2) Another solution is, after fetching the notebook, retrieve all the events of the notebook's namespace and filter. But this list can be huge...

**Comment by kimwnasptd** [comment]
@fediazgon I think we could try and implement a little bit of both. 

Regarding the `watch` of Events, lets 
1. Only add Events, to the controller's cache, that refer to a StatefulSet
2. Trigger the Reconcile function (with the mapFn) from Events, but lets use the Event's StatefulSet Name and Namespace as the Object/Key.

So if the Event was for a StatefulSet that corresponds to a Notebook, then the Reconcile function will continue. Else, the Reconcile function will just return. Then, we can list the Controller's watched Events and, if needed, update the condition from the relative Event. 

Regarding when to update the CR's `conditiona`. We need this logic to update the status of the Notebook CR only when the Pod is not submsitted to the API. This means that the `conditions/containerStatus` in the Notebook CR will both be empty. If the Pod gets POSTed to the API from the StatefulSet, then the CR's status would be updated from the Pod.

So we need to only list/filter the Events for a Notebook if
1. The underlying Pod was not found, and
2. The `conditions` and the `containerStatus` in that Notebook's CR are not set

I think this way we won't be calling the list/watch a lot and we will have a relatively confined list of watched events.

@fediazgon @jlewi WDYT?

**Comment by jlewi** [comment]
Do we need to get events just to determine that the pod hasn't started yet? Does the statefulset provide sufficient information about whether the part is started or not?

If we want the UI to provide detailed information about the reasons; should the UI just fetch the events?

**Comment by kimwnasptd** [comment]
If the StatefulSet can't create the Pod, then the only information it exposes is `replicas: 0` in its status field, which is not very enlightening regarding what went wrong. The information for why the Pod was not created exists in the events of the StatefulSet.

[ Just for clarification, when I say create the Pod, I mean that the STS POSTs the Pod to the API Server, not if the Pod is ready. ]

This seems to be the case only with StatefuleSet and not with Deployment for example. If we create a deployment and set a nonexistent Service Account for the Pod, then the deployment will surface the error information in its status while the StatefulSet will not.

> If we want the UI to provide detailed information about the reasons; should the UI just fetch the events?

I think the Notebooks Controller should be responsible to depict all the necessary information and status, of the underlying resources it creates, on the Custom Resource Instance.

**Comment by jlewi** [comment]
@kimwnasptd If events should be captured and turned into status/conditions information then why isn't stateful set doing that? If its a bug/oversight then it makes sense for us to fix that. If we capture events from the pod and turn that into notebook status are we doing something that is not best practice and or brittle?

In particular, if notebook controller is going around the statefulset API and looking at pod events then I think the notebook controller is making some assumptions about how statefulsets's work. 

Question

* If a pod doesn't start; what does the stateful set show?
* In Kubernetes to figure out why something didn't work; e.g. a CronJob failed you typically have to
  dig through the layers to get fine information; e.g. go from CronJob to Job to Pod
* Should we limit Notebook controller status and events to be functions of statefulset controller and events so we don't break the abstraction provided by statefulsets?
* How could we make it easy for users to navigate down the hierarchy from Notebook Resource -> statefulset -> pods
   
   * What if the Jupyter web app had a customizable way to set a link to the statefulset? 
   * There are lots of different Kubernetes monitoring solutions that would show users a list of resources (e.g. Statefulsets) and then allow people to navigate down to pods, logs and events
   * But the link would be different on different platforms. (e.g. some people might use DataDog or similar)
   * But if the web app could display a link for the statefulset then it would be easy for folks to take advantage of their monitoring solution to debug.

**Comment by kimwnasptd** [comment]
@jlewi 
> If a pod doesn't start; what does the stateful set show?

Only how many replicas the StatefulSet created and the number of Pods created by the StatefulSet controller that have a Ready Condition. It won't show any more info on its Status field regarding the underlying Pod's statuses.

> Should we limit Notebook controller status and events to be functions of statefulset controller and events so we don't break the abstraction provided by statefulsets?

This wouldn't be enough. If the StatefulSet creates the Pods, but the Pods can't be scheduled (insufficient cpu for example) then this information is not depicted neither in the StatefulSet's Events nor Status. 

This is why we need to make assumptions regarding how the StatefulSets work, because they don't fully expose the underlying state which we want to be able to show to the user.

> How could we make it easy for users to navigate down the hierarchy from Notebook Resource -> statefulset -> pods

Ideally, I think we should expose the less k8s constructs possible to the users/Data Scientists. If something would go wrong with their Notebook, then we should surface the details of the problem and not expect them to debug k8s. 

That's why I think we should be fully exposing the state of the underlying Resources, a Notebook CR has, to the user on their Notebook.

> But if the web app could display a link for the statefulset then it would be easy for folks to take advantage of their monitoring solution to debug.

This is a very interesting idea! We could have a field in the webapp's configmap with a parameterizable string that would link to a StatefulSet. Then the UI could expose that link and use the StatefulSet's info to show the correct link.

**Comment by jlewi** [comment]
So it seems like we have two possible paths forward

1. Try to aggregate information from sub-resources (e.g. Pod status, Pod events, etc...) into Notebook resource
1. In the UI make it easy to navigate to sub resources in existing Kubernetes UIs for monitoring

Thoughts on how we should proceed?

**Comment by fediazgon** [comment]
Even if we want to implement 2), we still need 1) to inform the user that something went wrong (stopping the spinner in the UI). Anyway, the solution that @kimwnasptd [proposed earlier](https://github.com/kubeflow/kubeflow/issues/3772#issuecomment-519980054) seems acceptable to me. We could start by adding that logic in the controller in one PR, and create another PR to reflect the information in the UI. Regarding how to proceed with 1).

* These are the logged events when you delete the "default-editor" service account and try to create a notebook server:

```text
8s          Normal    ProvisioningSucceeded   persistentvolumeclaim/workspace-test-nb-server                   Successfully provisioned volume pvc-903a0273-6606-49be-b3a8-3a2e83997f47
8s          Normal    ExternalProvisioning    persistentvolumeclaim/workspace-test-nb-server                   waiting for a volume to be created, either by external provisioner "k8s.io/minikube-hostpath" or manually created by system administrator
8s          Normal    Provisioning            persistentvolumeclaim/workspace-test-nb-server                   External provisioner is provisioning volume for claim "kubeflow/workspace-test-nb-server"
2s          Warning   FailedCreate            statefulset/test-nb-server                                       create Pod test-nb-server-0 in StatefulSet test-nb-server failed error: pods "test-nb-server-0" is forbidden: error looking up service account kubeflow/default-editor: serviceaccount "default-editor" not found
```
You can see how, in this case, the error surfaces to the statefulset.

* However, these are the events when you try to create a notebook server requesting more CPU units than available.

```text
37s         Warning   FailedScheduling        pod/test-nb-server-high-cpu-0                                   0/1 nodes are available: 1 Insufficient cpu.
37s         Normal    SuccessfulCreate        statefulset/test-nb-server-high-cpu                             create Pod test-nb-server-high-cpu-0 in StatefulSet test-nb-server-high-cpu successful
```
In this case, the statefulset does not surface the error from the pod that could not be scheduled. I.e., notebook-controller status cannot be function of statefulset's events, only of pod events. It can't be a function of the pod resource itself because the pod will be forever "pending":

```text
test-nb-server-high-cpu-0                                   0/1     Pending   0          14s
```

* These are the events after successfully creating  a notebook-server:

```text
8s          Normal    Scheduled               pod/test-nb-server-ok-0                                         Successfully assigned kubeflow/test-nb-server-ok-0 to minikube
8s          Normal    SuccessfulCreate        statefulset/test-nb-server-ok                                   create Pod test-nb-server-ok-0 in StatefulSet test-nb-server-ok successful
8s          Normal    ExternalProvisioning    persistentvolumeclaim/workspace-test-nb-server-ok               waiting for a volume to be created, either by external provisioner "k8s.io/minikube-hostpath" or manually created by system administrator
8s          Normal    Provisioning            persistentvolumeclaim/workspace-test-nb-server-ok               External provisioner is provisioning volume for claim "kubeflow/workspace-test-nb-server-ok"
8s          Normal    ProvisioningSucceeded   persistentvolumeclaim/workspace-test-nb-server-ok               Successfully provisioned volume pvc-2a42b7fd-cc26-485d-b991-2a90324cca11
6s          Normal    Created                 pod/test-nb-server-ok-0                                         Created container test-nb-server-ok
6s          Normal    Pulled                  pod/test-nb-server-ok-0                                         Container image "gcr.io/kubeflow-images-public/tensorflow-1.13.1-notebook-cpu:v0.5.0" already present on machine
6s          Normal    Started                 pod/test-nb-server-ok-0                                         Started container test-nb-server-ok
```

I suggest that instead of watching both pod resources (see [this code](https://github.com/kubeflow/kubeflow/blob/5c4fc0b0afab30b7a43dae534db8d3a633ccf214/components/notebook-controller/pkg/controller/notebook/notebook_controller.go#L137) and [this code](https://github.com/kubeflow/kubeflow/blob/5c4fc0b0afab30b7a43dae534db8d3a633ccf214/components/notebook-controller/pkg/controller/notebook/notebook_controller.go#L263)) and pod events (to solve the current issue). We watch pod events only, and we try to reconcile both cases (i.e., ok vs failed) using the information from those events (Normal Started vs Warning Failed Created).

Is that acceptable? Does someone know if a similar logic (watching events) is implemented in any of the [kubernetes/kubernetes controllers](https://github.com/kubernetes/kubernetes/tree/master/pkg/controller)?

**Comment by jlewi** [comment]
> We watch pod events only, and we try to reconcile both cases (i.e., ok vs failed) using the information from those events (Normal Started vs Warning Failed Created).

What does reconcile in this context mean? Do you mean the controller would take some action based on the pod events?

I think there are two questions here

1) Whether event information should be converted into status?
2) How do we propogate information about pods up to the UI?

Is turning events into status information a good pattern? K8s resources are supposed to be level based not edge triggered. If we depend on capturing events then it feels like we are violating that assumption.

If we follow a level based pattern for a notebook controller then my expectation would be that the controller could construct the correct status just by observing the current state of the world. If the status depends on capturing past events (which are ephemeral) then that seems to violate that assumption.

So I think if information is captured by events and not in status then it should probably stay as events.

Could the notebook controller reissue pod level events as notebook events? i.e. the notebook controller would watch for events associated with the pod and/or statefulset and then turn those into notebook related events?

Then the UI could just issue a get events for the notebook resource and surface them in the UI?
I believe the events would also be surfaced when calling `kubectl describe notebook ${NOTEBOOK}`.

**Comment by fediazgon** [comment]
> What does reconcile in this context mean? Do you mean the controller would take some action based on the pod events?

Yes, but

> K8s resources are supposed to be level based not edge triggered. If we depend on capturing events then it feels like we are violating that assumption.

Good point. Then I agree to turn pod events into notebook events. However, there are some statefulset events that also capture relevant information (see my first example earlier). I guess we should convert those too. 

However, right now, the notebook-controller watches Pod resources and turns them into notebook status. Should we keep this logic? Since now the UI is responsible for inferring the status by watching events... Also, if we decide to remove it, who is updating the [`NotebookStatus`]?(https://github.com/kubeflow/kubeflow/blob/5c4fc0b0afab30b7a43dae534db8d3a633ccf214/components/notebook-controller/pkg/apis/notebook/v1alpha1/notebook_types.go#L41)

About 

> Whether event information should be converted into status?

@kimwnasptd?

**Comment by kimwnasptd** [comment]
@jlewi @fediazgon I also totally agree with watching the Pod/StatefulSet events and propagating them as Notebook Events. We can then add a `View logs` button to the UI for each Notebook to make it easier to debug and monitor each instance.

Regarding the Status, my proposal would be to mirror the Status of the underlying Pod. This would include all the `conditions` and `containerStatuses` of all the Containers (might need to extend the API for this). But we can discuss this in a separate issue.

So to sum up this conversation. For this issue, lets have the Notebooks Controller create Notebook Events by propagating both StatefulSet and Pod Events. This will help users debug their Notebooks by running just `kubectl describe` on the Notebook, as Jeremy pointed out.

With this established, I can then move forward and add the `View logs` button to the UI.

@jlewi @fediazgon does this sound good to you?

**Comment by lluunn** [comment]
> This seems to be the case only with StatefuleSet and not with Deployment for example. If we create a deployment and set a nonexistent Service Account for the Pod, then the deployment will surface the error information in its status while the StatefulSet will not.

Is this intended, or is something StatefulSet should fix?

**Comment by kimwnasptd** [comment]
@lluunn I am not sure. I've tried searching the kubernetes docs/API Specs but can't find the rationale behind it.

**Comment by jlewi** [comment]
> So to sum up this conversation. For this issue, lets have the Notebooks Controller create Notebook Events by propagating both StatefulSet and Pod Events. This will help users debug their Notebooks by running just kubectl describe on the Notebook, as Jeremy pointed out.

@kimwnasptd LGTM; thanks!

**Comment by jlewi** [comment]
@kimwnasptd Any update on this?

**Comment by fediazgon** [comment]
I'm working on the non-UI part. I know that is taking me a while but I've been catching up of lectures because I want to understand k8s programming principles before writing code.

If it is urgent, someone can take it, otherwise, I would like to keep it.

**Comment by fediazgon** [comment]
Please @kunmingg, @jlewi, see PRs above. On second thoughts, I don't like the idea of the UI surfacing events as status. Events have a ttl, which means that, if the web-app can't fetch any nb-related event, it must assume that the nb is running (see [this](https://github.com/kubeflow/kubeflow/pull/4214/files#diff-d575d8cda5f10ed63f52ac8e7a65b733R258)). Also, the user might delete all events via `kubectl` and this should not change the real status of the notebook.

I'd rather parse pod and sts events as API NotebookStatus in the controller, and avoid delegating this responsibility in the web-app.

**Comment by jlewi** [comment]
@fediazgon Turning events into status doesn't like a common pattern in Kubernetes for the reasons mentioned previously.

> Events have a ttl, which means that, if the web-app can't fetch any nb-related event, it must assume that the nb is running (see this). 

Events have a TTL but the controller should continue emitting events. 

Reconcile is invoked periodically. So even hours after a notebook starts, reconcile should be invoked. The controller should check the status of the world and emit appropriate events.

So for example, if the notebook controller is waiting for the pod to start then on every invocation of reconcile it could issue an event e.g. "Waiting for notebook to start" and as mentioned previously that event could potentially contain additional information such as "Waiting for notebook to start; missing volume ..."

**Comment by fediazgon** [comment]
Would that imply sending an "Everything is OK" event periodically? I'm concerned about the two following flows:

1) User creates nb in the dashboard -> UI immediately fetches nb events -> No events yet (Real status is "waiting").
1) Nb is running -> Reconcile stops reissuing events cause everything is ok ->  UI fetches nb events -> No events (Real status is "running").

**Comment by jlewi** [comment]
@fediazgon You raise a good point. Wouldn't this be covered by conditions?

* The UI should use conditions on the Jupyter notebook to determine overall status of the notebook

   * E.g. the Jupyter notebook controller should define appropriate conditions Ready or NotReady
   * And the last condition should be continually updated by the controller based on whether the notebook is actually ready or note

      * So I think this issue addresses your use case

* Surfacing events is about providing the user with detailed information about why the notebook is starting; e.g. because the pod doesn't exist

      * The retention of events in Kubernetes should be much longer than the reconcile period
         * e.g. reconcile is order of seconds to minutes; events retention is on the order of hours

      * So if a notebook is stuck in a NotReady state and we fetch the last 2 hours of events for the notebook I would expect we would see events indicating why it failed to start.

**Comment by jlewi** [comment]
@fediazgon @lluunn @kimwnasptd it looks like #4139 was merged.  

What are the next steps? Is there more work to be done in the controller to surface the information needed by the UI? Or is the remaining work at this point just to update the UI to surface the information #3637?

**Comment by fediazgon** [comment]
UI work only (https://github.com/kubeflow/kubeflow/pull/4214) + change manifests to allow the controller to get/create/watch events.

**Comment by jlewi** [comment]
It looks like kubeflow/manifests#641 was merged. So I guess the only remaining work is the jupyter web app changes which is
kubeflow/kubeflow#3637 and looks like there is an open PR kubeflow/kubeflow#4214.

I think we can close out this issue; before we do @fediazgon would it be possible for verification purposes just to past into the issue the output of `kubectl describe notebooks ${SOMENOTEBOOK}` just to document that the pod events are being surfaced by the notebook.

**Comment by fediazgon** [comment]
> kubectl describe notebooks ${SOMENOTEBOOK}

For a healthy and non-healthy nb:

```
  ~ ❯ kubectl describe notebooks ok-nb -n fediazgon                                                                                                                                                                                                                                                                                            at ⎈ kf-3372 at  12:10:16
Name:         ok-nb
Namespace:    fediazgon
Labels:       app=ok-nb
Annotations:  <none>
API Version:  kubeflow.org/v1beta1
Kind:         Notebook
Metadata:
  Creation Timestamp:  2020-01-08T11:10:12Z
  Generation:          1
  Resource Version:    26737
  Self Link:           /apis/kubeflow.org/v1beta1/namespaces/fediazgon/notebooks/ok-nb
  UID:                 7c3030b3-d056-437c-a11b-3e8adfe86918
Spec:
  Template:
    Spec:
      Containers:
        Env:
        Image:  gcr.io/kubeflow-images-public/tensorflow-1.14.0-notebook-cpu:v0.7.0
        Name:   ok-nb
        Resources:
          Limits:
          Requests:
            Cpu:     0.5
            Memory:  1.0Gi
        Volume Mounts:
          Mount Path:              /home/jovyan
          Name:                    workspace-ok-nb
          Mount Path:              /dev/shm
          Name:                    dshm
      Service Account Name:        default-editor
      Ttl Seconds After Finished:  300
      Volumes:
        Name:  workspace-ok-nb
        Persistent Volume Claim:
          Claim Name:  workspace-ok-nb
        Empty Dir:
          Medium:  Memory
        Name:      dshm
Status:
  Conditions:
    Last Probe Time:  2020-01-08T11:10:16Z
    Type:             Running
    Last Probe Time:  2020-01-08T11:10:14Z
    Reason:           PodInitializing
    Type:             Waiting
  Container State:
    Running:
      Started At:  2020-01-08T11:10:16Z
  Ready Replicas:  1
Events:
  Type     Reason            Age   From                 Message
  ----     ------            ----  ----                 -------
  Normal   Scheduled         29s   notebook-controller  Reissued from pod/ok-nb-0: Successfully assigned fediazgon/ok-nb-0 to minikube
  Normal   SuccessfulCreate  29s   notebook-controller  Reissued from statefulset/ok-nb: create Pod ok-nb-0 in StatefulSet ok-nb successful
  Normal   Pulled            29s   notebook-controller  Reissued from pod/ok-nb-0: Container image "docker.io/istio/proxy_init:1.1.6" already present on machine
  Normal   Created           29s   notebook-controller  Reissued from pod/ok-nb-0: Created container istio-init
  Normal   Started           29s   notebook-controller  Reissued from pod/ok-nb-0: Started container istio-init
  Normal   Pulled            28s   notebook-controller  Reissued from pod/ok-nb-0: Container image "gcr.io/kubeflow-images-public/tensorflow-1.14.0-notebook-cpu:v0.7.0" already present on machine
  Normal   Created           28s   notebook-controller  Reissued from pod/ok-nb-0: Created container ok-nb
  Normal   Started           28s   notebook-controller  Reissued from pod/ok-nb-0: Started container ok-nb
  Normal   Pulled            28s   notebook-controller  Reissued from pod/ok-nb-0: Container image "docker.io/istio/proxyv2:1.1.6" already present on machine
  Normal   Created           28s   notebook-controller  Reissued from pod/ok-nb-0: Created container istio-proxy
  Normal   Started           27s   notebook-controller  Reissued from pod/ok-nb-0: Started container istio-proxy
  Warning  Unhealthy         25s   notebook-controller  Reissued from pod/ok-nb-0: Readiness probe failed: HTTP probe failed with statuscode: 503
  ~ ❯ kubectl describe notebooks not-enough-cpu-nb -n fediazgon                                                                                                                                                                                                                                                                                at ⎈ kf-3372 at  12:10:43
Name:         not-enough-cpu-nb
Namespace:    fediazgon
Labels:       app=not-enough-cpu-nb
Annotations:  <none>
API Version:  kubeflow.org/v1beta1
Kind:         Notebook
Metadata:
  Creation Timestamp:  2020-01-08T11:11:11Z
  Generation:          1
  Resource Version:    26831
  Self Link:           /apis/kubeflow.org/v1beta1/namespaces/fediazgon/notebooks/not-enough-cpu-nb
  UID:                 a45c81a3-ef9c-4838-9e95-7499ecf0990c
Spec:
  Template:
    Spec:
      Containers:
        Env:
        Image:  gcr.io/kubeflow-images-public/tensorflow-1.14.0-notebook-cpu:v0.7.0
        Name:   not-enough-cpu-nb
        Resources:
          Limits:
          Requests:
            Cpu:     2
            Memory:  1.0Gi
        Volume Mounts:
          Mount Path:              /home/jovyan
          Name:                    workspace-not-enough-cpu-nb
          Mount Path:              /dev/shm
          Name:                    dshm
      Service Account Name:        default-editor
      Ttl Seconds After Finished:  300
      Volumes:
        Name:  workspace-not-enough-cpu-nb
        Persistent Volume Claim:
          Claim Name:  workspace-not-enough-cpu-nb
        Empty Dir:
          Medium:  Memory
        Name:      dshm
Events:
  Type     Reason            Age   From                 Message
  ----     ------            ----  ----                 -------
  Normal   SuccessfulCreate  11s   notebook-controller  Reissued from statefulset/not-enough-cpu-nb: create Pod not-enough-cpu-nb-0 in StatefulSet not-enough-cpu-nb successful
  Warning  FailedScheduling  11s   notebook-controller  Reissued from pod/not-enough-cpu-nb-0: 0/1 nodes are available: 1 Insufficient cpu.
```

**Comment by jlewi** [comment]
Thanks @fediazgon

**Comment by lalithvaka** [comment]
@jlewi , @fediazgon I ran into this issue and I do not see any events populating on my UI even though I have couple of notebooks in forever spinning state. Not sure if I am missing anything? Am running KF 1.0.2 version.  Attaching a screenshot on the behavior. Shall I create a new issue for this?
![image](https://user-images.githubusercontent.com/8292387/82748587-2617d380-9d58-11ea-8337-7788655ab359.png)
